### PR TITLE
[DOCFIX] Add global state section to unit testing guide

### PR DESCRIPTION
**Pull request by aaudiber**
image

### CODE REVIEW

File: docs/en/Unit-Testing.md [hunk 1]
```diff
@@ -84,6 +84,33 @@ public void detectLostWorker() throws Exception {
...
-5. Avoid slow tests. Mock expensive dependencies and aim to keep individual test times under 100ms.
+3. Avoid using Whitebox to mess with the internal state of objects under test. If you need to mock a dependency, change the object to take the dependency as a parameter in its constructor (see [dependency injection](https://en.wikipedia.org/wiki/Dependency_injection))
+4. Avoid slow tests. Mock expensive dependencies and aim to keep individual test times under 100ms.
+
+# Managing Global State
+All tests in a project run in the same JVM, so it's important to properly manage JVM-global state so that tests don't interfere with each other. Our solution to this issue is to use JUnit `@Rule` to manage global state.
+
+### Changing Alluxio configuration during a test
+Some unit tests want to test Alluxio under different configurations. This requires modifying the global `Configuration` object. When all tests in a suite need configuration parameters set a certain way, use `ConfigurationRule` to set them. For configuration changes needed for an individual test, use `Configuration.set(key, value`, and create an `@After` method to clean up the configuration change after the test:
```
Comment (apc999) [inline docs/en/Unit-Testing.md]: missing right Parentheses  in `Configuration.set(key, value`

File: docs/en/Unit-Testing.md [hunk 2]
```diff
@@ -84,6 +84,33 @@ public void detectLostWorker() throws Exception {
...
-4. Avoid using Whitebox to mess with the internal state of objects under test. If you need to mock a dependency, change the object to take the dependency as a parameter in its constructor (see [dependency injection](https://en.wikipedia.org/wiki/Dependency_injection))
-5. Avoid slow tests. Mock expensive dependencies and aim to keep individual test times under 100ms.
+3. Avoid using Whitebox to mess with the internal state of objects under test. If you need to mock a dependency, change the object to take the dependency as a parameter in its constructor (see [dependency injection](https://en.wikipedia.org/wiki/Dependency_injection))
+4. Avoid slow tests. Mock expensive dependencies and aim to keep individual test times under 100ms.
+
+# Managing Global State
+All tests in a project run in the same JVM, so it's important to properly manage JVM-global state so that tests don't interfere with each other. Our solution to this issue is to use JUnit `@Rule` to manage global state.
```
Comment (apc999) [inline docs/en/Unit-Testing.md]: "All tests in a project run in the same JVM" --- even for unit tests in different modules?

Comment (apc999) [inline docs/en/Unit-Testing.md]: Can we first give a few examples of what we mean by "JVM-global state" in the beginning of this section?

File: docs/en/Unit-Testing.md [hunk 3]
```diff
@@ -84,6 +84,33 @@ public void detectLostWorker() throws Exception {
...
+
+### Changing Alluxio configuration during a test
+Some unit tests want to test Alluxio under different configurations. This requires modifying the global `Configuration` object. When all tests in a suite need configuration parameters set a certain way, use `ConfigurationRule` to set them. For configuration changes needed for an individual test, use `Configuration.set(key, value`, and create an `@After` method to clean up the configuration change after the test:
+
+```java
+@After
+public void after() {
+  ConfigurationTestUtils.resetConfiguration();
```
Comment (apc999) [inline docs/en/Unit-Testing.md]: actually, if this is frequent, shall we make this into `TestUtils`?

File: docs/en/Unit-Testing.md [hunk 4]
```diff
@@ -84,6 +84,33 @@ public void detectLostWorker() throws Exception {
...
+}
+```
+
+### Changing System properties during a test
```
Comment (apc999) [inline docs/en/Unit-Testing.md]: shall we also have  a subsection talking about changing system properties for a test suit?
```java
  @ClassRule
  public static SystemPropertyRule sDisableHdfsCacheRule =
      new SystemPropertyRule("fs.hdfs.impl.disable.cache", "true");
```

File: docs/en/Unit-Testing.md [hunk 5]
```diff
@@ -84,6 +84,33 @@ public void detectLostWorker() throws Exception {
...
-4. Avoid using Whitebox to mess with the internal state of objects under test. If you need to mock a dependency, change the object to take the dependency as a parameter in its constructor (see [dependency injection](https://en.wikipedia.org/wiki/Dependency_injection))
-5. Avoid slow tests. Mock expensive dependencies and aim to keep individual test times under 100ms.
+3. Avoid using Whitebox to mess with the internal state of objects under test. If you need to mock a dependency, change the object to take the dependency as a parameter in its constructor (see [dependency injection](https://en.wikipedia.org/wiki/Dependency_injection))
+4. Avoid slow tests. Mock expensive dependencies and aim to keep individual test times under 100ms.
+
+# Managing Global State
+All tests in a project run in the same JVM, so it's important to properly manage JVM-global state so that tests don't interfere with each other. Our solution to this issue is to use JUnit `@Rule` to manage global state.
+
+### Changing Alluxio configuration during a test
```
Comment (apc999) [inline docs/en/Unit-Testing.md]: Having a section talking about `ConfigurationRule`?
```java
  @Rule
  public ConfigurationRule mConfigurationRule = new ConfigurationRule(ImmutableMap.of(
       PropertyKey.USER_UFS_DELEGATION_READ_BUFFER_SIZE_BYTES, Long.toString(FILE_SIZE)));
```

**Comment by apc999** [comment]
@aaudiber I left a few comments here.

File: docs/en/Unit-Testing.md (hunk 5 above)
Comment (gpang) [inline docs/en/Unit-Testing.md]: I think there should be sub sections for the different methods. configuration for test suite, configuration for test case, system property for test suite, system property for test case.

**Comment by gpang** [comment]
@aaudiber Thanks for this great improvement! I left a suggestion.

**Comment by apc999** [comment]
@aaudiber  is this PR still active? thanks

File: docs/en/Unit-Testing.md (hunk 1 above)
Comment (aaudiber) [inline docs/en/Unit-Testing.md]: fixed, thanks

File: docs/en/Unit-Testing.md (hunk 2 above)
Comment (aaudiber) [inline docs/en/Unit-Testing.md]: Sorry, I meant module, updated

I've added examples of JVM-global state

File: docs/en/Unit-Testing.md (hunk 3 above)
Comment (aaudiber) [inline docs/en/Unit-Testing.md]: You mean rename the class to `TestUtils`? I don't have a preference towards either name, so I'd be fine with that.

File: docs/en/Unit-Testing.md (hunk 4 above)
Comment (aaudiber) [inline docs/en/Unit-Testing.md]: added

File: docs/en/Unit-Testing.md (hunk 5 above)
Comment (aaudiber) [inline docs/en/Unit-Testing.md]: added examples for all four cases

**Comment by aaudiber** [comment]
@apc999 Sorry for the delay, I've addressed the comments now

File: docs/en/Unit-Testing.md (hunk 3 above)
Comment (apc999) [inline docs/en/Unit-Testing.md]: I think I think this is out of the scope of this PR.  We could discuss later

**Comment by apc999** [comment]
LGTM. thanks for the improvement!

**Comment by gpang** [comment]
LGTM

**Pull request merged**